sdk/evaluation/azure-ai-evaluation/azure/ai/evaluation/_evaluators/_document_retrieval/_document_retrieval.py

```diff
@@ -0,0 +1,323 @@
+        """
+        # Set the scoring function
+        def calculate_dcg(relevance: float, rank: int):
+            return ((math.pow(2, relevance) - 1) / (math.log2(rank + 1)))
+
+        ranks = list(range(1, self.k + 1))
+        dcg = sum(starmap(calculate_dcg, zip(result_docs_groundtruth_labels, ranks)))
+        idcg = sum(starmap(calculate_dcg, zip(ideal_docs_groundtruth_labels, ranks)))
+        ndcg = dcg / float(idcg)
+
+        return ndcg
+    
+    def _compute_xdcg(
+        self,
+        result_docs_groundtruth_labels: List[float]
+    ) -> float:
+        """
+        XDCG calculated for the top 3 documents retrieved from a search query.
+        XDCG measures how objectively good are the top 3 documents, discounted by their position in the list.
+        """
+        def calculate_xdcg_numerator(relevance, rank):
+            return (25 * relevance * math.pow(self.xdcg_discount_factor, rank - 1))
+        
+        def calculate_xdcg_denominator(rank):
+            return math.pow(self.xdcg_discount_factor, rank - 1)
+
+        ranks = list(range(1, self.k + 1))
+        xdcg_n = sum(starmap(calculate_xdcg_numerator, zip(result_docs_groundtruth_labels, ranks)))
+        xdcg_d = sum(map(calculate_xdcg_denominator, ranks))
+
+        return xdcg_n / float(xdcg_d)
+    
+    def _compute_fidelity(
+        self,
+        result_docs_groundtruth_labels: List[float],
+        ideal_docs_groundtruth_labels: List[float],
+    ) -> float:
+        """
+        Fidelity calculated over all documents retrieved from a search query.
+        Fidelity measures how objectively good are all of the documents retrieved compared with all known good documents in the underlying data store.
+        """
+        def calculate_weighted_sum_by_rating(labels: List[float]) -> float:
+            s = self.groundtruth_min + self.groundtruth_step
+
+            # get a count of each label
+            label_counts = {str(i): 0 for i in range(s, self.groundtruth_max + 1, self.groundtruth_step)}
+
+            for label in labels:
+                if label >= s:
+                    label_counts[str(label)] += 1
+
+            sorted_label_counts = [x[1] for x in sorted(label_counts.items(), key=lambda x: x[0])]
+
+            # calculate weights
+            weights = [(math.pow(2, i+1) - 1) for i in range(s, self.groundtruth_max + 1, self.groundtruth_step)]
+
+            # return weighted sum
+            return sum(starmap(operator.mul, zip(sorted_label_counts, weights)))
+        
+        weighted_sum_by_rating_results = calculate_weighted_sum_by_rating(result_docs_groundtruth_labels)
+        weighted_sum_by_rating_index = calculate_weighted_sum_by_rating(ideal_docs_groundtruth_labels)
+
+        if weighted_sum_by_rating_index == 0:
+            return math.nan
+
+        return weighted_sum_by_rating_results / float(weighted_sum_by_rating_index)
+            
+
+    def _compute_fidelity_old(
```

Comment:
This is the formulation depending on a static 0-4 scale for qrels.  I'll remove it later, since I've replaced it with a more generalized formula above.

Reply:
Removed this.